### PR TITLE
Core credentials API refactoring

### PR DESCRIPTION
**Pull request by jboeuf**
As part of having all the gRPC libraries be consistent, we would like to make a breaking change to the current credentials API.

# Motivation

The idea coming from @murgatroid99 is that there are now 2 credentials classes, one for channels (grpc_channel_credentials) to be set on a channel and the other one on calls (grpc_call_credentials). We still have composition functions that do:
`channel_creds <- compose(channel_creds, call_creds)`
`call_creds <- compose(call_creds, call_creds)`

This refactoring will achieve 2 main goals:

1. Add some type safety so that users of the API do not set call credentials only (without a channel one) when creating a channel. Likewise, this will make it impossible to set channel creds on a particular call which is currently just a runtime error.

2. This will also cleanup a lot of the implementation that basically had to provide at runtime for what the type safety was not enforcing.

# API breakage

In terms of API breakage, this translate into the following changes:

### Objects
 `grpc_credentials` is now replaced by `grpc_channel_credentials` for channels and `grpc_call_credentials` for calls.

### Factory Functions 
`grpc_ssl_credentials(...)` and `grpc_google_default_credentials(...)` return a `grpc_channel_credentials *` and the rest of the credentials factories exposed in `grpc_security.h` return a `grpc_call_credentials *`.

For composition we now have the following functions instead of `grpc_composite_credentials_create`:
```c
grpc_channel_credentials *grpc_composite_channel_credentials_create(
    grpc_channel_credentials *channel_creds, grpc_call_credentials *call_creds);

grpc_call_credentials *grpc_composite_call_credentials_create(
    grpc_call_credentials *creds1, grpc_call_credentials *creds2);
```

### Objects
`credentials.h` now contains `ChannelCredentials` and `CallCredentials` respectively for channel and client credentials. 

### Factory Functions
`SslCredentials(...)` , `GoogleDefaultCredentials(...)` and `InsecureChannelCredentials(...)` now return a `std::shared_ptr<ChannelCredentials>`. The rest of the factory functions return a `std::shared_ptr<CallCredentials>`.

For composition we now have the following functions instead of `CompositeCredentials()`:
```cpp
std::shared_ptr<ChannelCredentials> CompositeChannelCredentials(
    const std::shared_ptr<ChannelCredentials>& channel_creds,
    const std::shared_ptr<CallCredentials>& call_creds);

std::shared_ptr<CallCredentials> CompositeCalllCredentials(
    const std::shared_ptr<CallCredentials>& creds1,
    const std::shared_ptr<CallCredentials>& creds2);
```

Note that `InsecureCredentials()` has been renamed `InsecureChannelCredentials()` and we foresee (based on our changes in the tests) this particular API breakage to be the most painful for our user base.

### CODE REVIEW

**Comment by jboeuf** [comment]
@jtattermusch @murgatroid99 do you want to have a quick look? I'm pretty sure that this is aligned with what we discussed but I wanted to give you an early preview.

**Comment by jboeuf** [comment]
@ctiller @a11r @jayantkolhe @yang-g @nathanielmanistaatgoogle @tbetbetbe @nicolasnoble @dgquintas @iamqizhao please let me know if you have a some feedback regarding this proposal.

**Comment by nathanielmanistaatgoogle** [comment]
@soltanmm may also want to look at this.

I like the change content but let me ask one contrarian process question: must the old methods be removed in the same pull request in which the new methods are introduced? The names involved seem non-overlapping; why not introduce the new methods in a small pull request (and mark the old ones deprecated) and then remove the old ones later? Aren't many small pulls request better than one large one?

**Comment by jboeuf** [comment]
I would be great if we could do that. However the names are indeed very
much overlapping: all the credentials factory functions now return another
type (grpc_channel_credentials/grpc_call_credentials as opposed to
grpc_credentials). The channel and call constructors have also changed
signature. It would be too much work to support both IMO and i'd much be
done with everything at once since the changes should be very much
mechanical.

I'm happy to fix the python layer and other languages as well if people are
too busy. I'd much rather this change to be completely atomic.

On Mon, Oct 12, 2015 at 3:26 PM, Nathaniel Manista <notifications@github.com

**Comment by yang-g** [comment]
LGTM
Could you also fix the code under examples/c++? Thanks.

**Comment by jboeuf** [comment]
@ctiller Can you please review the core API change and make sure that you're OK with it? I'll ask the owners of the different wrapped languages to adapt then (with my help if they need to of course).

Thanks @soltanmm for prompting this request.

**Comment by ctiller** [comment]
LGTM... please bump the minor version in build.yaml before submission.

**Comment by jboeuf** [comment]
Thanks @ctiller. I bumped the minor version as requested.

@soltanmm (and/or @nathanielmanistaatgoogle) , @jtattermusch, @tbetbetbe, @stanley-cheung, @jcanizales, @murgatroid99, can you please do some pull requests against this branch to adapt your respective wrapped language to the new API?
Thanks much! Let me know if you need help with that.

**Comment by jcanizales** [comment]
Maybe I'm not reading the motivation correctly, but IIUC this is just refactoring without changing functionality?

Since we entered beta, I think our bar for breaking the surface of the libraries has to be of the sort of "sorry, but we screwed up and didn't realize we couldn't support use case X with additions to our current surface". Otherwise, we have to make the extra effort of making the change backward-compatible, or just do it on a v2-alpha branch.

**Comment by jboeuf** [comment]
@jcanizales Yes, you're right that was the goal. I requested this exception to @jayantkolhe though because we had a lot of talks around usable auth with the other wrapped languages folks that occurred post-beta (and, in retrospect, I should definitely have included you, sorry about that). I truly think that this is a major improvement of our API surface and I'm happy to discuss it in person.

File: src/core/security/credentials.c [hunk 1]
```diff
@@ -1213,32 +1093,46 @@ grpc_credentials *grpc_credentials_contains_type(
...
+  if (channel_creds == NULL) return NULL;
+  if (channel_creds->call_creds == NULL) {
+    channel_creds->call_creds = grpc_call_credentials_ref(call_creds);
+  } else if (call_creds != NULL) {
+    grpc_call_credentials *composite_creds =
+        grpc_composite_call_credentials_create(channel_creds->call_creds,
+                                               call_creds, NULL);
+    if (composite_creds == NULL) return NULL;
+    grpc_call_credentials_unref(channel_creds->call_creds);
+    channel_creds->call_creds = composite_creds;
```
Comment (murgatroid99) [inline src/core/security/credentials.c]: Why does this modify the channel credentials in place? I very much expect composition to be idempotent.

Comment (jboeuf) [inline src/core/security/credentials.c]: That's a fair point. Let me have a look at this.

**Comment by jboeuf** [comment]
@murgatroid99 I addressed the issue that you raised. Please see b73cbc2

**Comment by jtattermusch** [comment]
I've migrated C# to the new core creds API: https://github.com/jboeuf/grpc/pull/6
@jboeuf  please take a look.

**Comment by jtattermusch** [comment]
@tbetbetbe , @soltanmm , @stanley-cheung, @murgatroid99 there's still a PR needed from each of you to migrate your languages to the new core API. Let's get this PR merged soon.

**Comment by murgatroid99** [comment]
I sent jboeuf/grpc#3 and stanley sent jboeuf/grpc#4.

**Comment by jtattermusch** [comment]
@murgatroid99  Ah sorry, I didn't see that. Thanks!

**Comment by tbetbetbe** [comment]
I just sent https://github.com/jboeuf/grpc/pull/7

**Comment by jboeuf** [comment]
Thanks everyone for the PRs. Looks like python is the only remaining guy. Assigning to @soltanmm. Please let me know if you need help.

**Comment by jboeuf** [comment]
@soltanmm @nathanielmanistaatgoogle any chance we can get python done this week? Thanks much!

**Comment by jtattermusch** [comment]
@soltanmm  escalating this to P0.

**Comment by thelinuxfoundation** [comment]
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* One or more commits are not associated with a GitHub user. The email used in the Git commit must match a verified email address in a GitHub user profile. For more information, please see https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/
* User @soltanmm isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

**Pull request merged**